File: src/leetCodeProblems/ArrayMatrixTwoD/FindWinnerOnTicTacToeGame1275.go
package main

import "log"

/*
- LeetCode - https://leetcode.com/problems/find-winner-on-a-tic-tac-toe-game/
- Space - O(m*n)
- Time - O(m*n)
*/

func winner(w int) string {
	switch w {
	case 1:
		return "A"
	case 2:
		return "B"
	default:
		return "None"
	}
}

func tictactoe(moves [][]int) string {
	board := make([][]int, 3)

	for i := range board {
		board[i] = make([]int, 3)
	}

	for i, v := range moves {
		if i%2 == 0 {
			board[v[0]][v[1]] = 1
		} else {
			board[v[0]][v[1]] = 2
		}
	}

	rowsCheck := checkRows(board)
	if rowsCheck != -1 {
		return winner(rowsCheck)
	}

	columnsCheck := checkColumns(board)
	log.Println(columnsCheck)
	if columnsCheck != -1 {
		return winner(columnsCheck)
	}

	diagonalsCheck := checkDiagonal(board)
	if diagonalsCheck != -1 {
		return winner(diagonalsCheck)
	}

	if len(moves) < 9 {
		return "Pending"
	}
	return "Draw"
}

func checkRows(g [][]int) int {
	for _, v := range g {
		if v[2] != 0 && v[0] == v[1] && v[1] == v[2] {
			return v[0]
		}
	}
	return -1
}
func checkColumns(g [][]int) int {
	for i := 0; i < 3; i++ {
		if g[0][i] != 0 && g[0][i] == g[1][i] && g[1][i] == g[2][i] {
			return g[0][i]
		}
	}
	return -1
}

func checkDiagonal(g [][]int) int {
	if g[0][0] != 0 && g[0][0] == g[1][1] && g[1][1] == g[2][2] {
		return g[0][0]
	}
	if g[2][0] != 0 && g[2][0] == g[1][1] && g[1][1] == g[0][2] {
		return g[2][0]
	}
	return -1
}

// func main() {
// 	// log.Println(tictactoe([][]int{{0, 0}, {2, 0}, {1, 1}, {2, 1}, {2, 2}}))
// 	// log.Println(tictactoe([][]int{{0, 0}, {1, 1}, {0, 1}, {0, 2}, {1, 0}, {2, 0}}))
// 	// log.Println(tictactoe([][]int{{0, 0}, {1, 1}, {2, 0}, {1, 0}, {1, 2}, {2, 1}, {0, 1}, {0, 2}, {2, 2}}))
// 	log.Println(tictactoe([][]int{{2, 0}, {1, 1}, {0, 2}, {2, 1}, {1, 2}, {1, 0}, {0, 0}, {0, 1}}))

// }

File: src/leetCodeProblems/Matrix/FindWinnerOnTicTacToeGame1275.java
package leetCodeProblems.ArrayMatrixTwoD;

/**
 * LeetCode - https://leetcode.com/problems/find-winner-on-a-tic-tac-toe-game/solution/
 *
 * TimeComplexity - O(m*n)
 * SpaceComplexity - O(n2)
 */
public class FindWinnerOnTicTacToeGame1275 {

    private String getWinnerName(int winner) {

        switch (winner) {
            case 0:
                return "NoWinner";
            case 1:
                return "A";
            case -1:
                return "B";
        }
        return "NoWinner";
    }

    private int checkDiagonal(int[][] board) {

        int currentPlayer = board[0][0];
        int currentPlayerWinnerCount = 1;

        for(int i=1; i < 3; i++) {

            if (board[i][i] == currentPlayer) {
                currentPlayerWinnerCount++;
            }
        }

        if (currentPlayerWinnerCount == 3) {
            return currentPlayer;
        }

        return 0;
    }

    private int checkAntiDiagonal(int[][] board) {

        int currentPlayer = board[0][2];

        if (board[1][1] == currentPlayer && board[2][0] == currentPlayer) {
            return currentPlayer;
        }

        return 0;
    }

    private int checkRows(int[][] board) {

        for(int i=0; i < 3; i++) {
            int currentPlayer = board[i][0];

            if (board[i][1] == currentPlayer && board[i][2] == currentPlayer) {
                return currentPlayer;
            }
        }

        return 0;
    }

    private int checkColumns(int[][] board) {

        for(int i=0; i < 3; i++) {
            int currentPlayer = board[0][i];

            if (board[1][i] == currentPlayer && board[2][i] == currentPlayer) {
                return currentPlayer;
            }
        }

        return 0;
    }

    public String tictactoe(int[][] moves) {

        int[][] board = new int[3][3];

        for(int i=0; i < moves.length; i++) {
            if (i%2 == 0) {
                board[moves[i][0]][moves[i][1]] = 1; //A player
            }
            else {
                board[moves[i][0]][moves[i][1]] = -1; //B player
            }
        }

        String diagonalWinner = getWinnerName(checkDiagonal(board));

        if (!diagonalWinner.equals("NoWinner")) {
            return diagonalWinner;
        }

        String antiDiagonalWinner = getWinnerName(checkAntiDiagonal(board));

        if (!antiDiagonalWinner.equals("NoWinner")) {
            return antiDiagonalWinner;
        }

        String rowWinner = getWinnerName(checkRows(board));

        if (!rowWinner.equals("NoWinner")) {
            return rowWinner;
        }

        String columnWinner = getWinnerName(checkColumns(board));

        if (!columnWinner.equals("NoWinner")) {
            return columnWinner;
        }

        if (moves.length < 9) {
            return "Pending";
        }

        return "Draw";
    }

    public static void main(String[] args) {
        //int[][] moves = {{0,0}, {1,1}, {2,0}, {1,0}, {1,2}, {2,1}, {0,1}, {0,2}, {2,2}};

        //int[][] moves = {{0,0}, {2,0}, {1,1}, {2,1}, {2,2}};

        //int[][] moves = {{0,0}, {1,1}, {0,1}, {0,2}, {1,0}, {2,0}};

        int[][] moves = {{1,2},{2,1},{1,0},{0,0},{0,1},{2,0},{1,1}};

        FindWinnerOnTicTacToeGame1275 obj = new FindWinnerOnTicTacToeGame1275();

        System.out.println(obj.tictactoe(moves));
    }
}
